### PR TITLE
fix(risedev): do not merge config with `risingwave.toml`

### LINKED ISSUE #6803: config: Use only one set of default values for risingwave.toml
Currently we have `config/risingwave.toml` (used by risedev -> used by ci) and default value defined in code. They did intentionally have some different values, e.g., https://github.com/risingwavelabs/risingwave/pull/6553. The reason is that we want to use different defaults for ci/dev & cloud.

https://github.com/risingwavelabs/risingwave/blob/533b9a966873ee06a754ca42a882e54e425845f3/src/config/README.md#L11-L15

This may be confusing and cause `risedev` need to merge `config-path` with the base one, which is a little bit hacky.

If we can only keep one set of defaults, things can be more straightforward. But I'm not sure how to decide the values.

### PR DESCRIPTION
**Pull request by fuyufjh**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

Previously in #6678, the user-specified `config-path` would be merged with default `risingwave.toml` implicitly. This behavior looks opaque to me and it may be a pitfall that makes the `risingwave.toml` become another copy of default parameters, beside the defaults in `RwConfig`. 

This PRs simplifies it: `risedev` just copies the specified `config-path` to `.risingwave/risingwave.toml` and then uses it as the config. No merging anymore.

If `config-path` is not specified in the profile, then `--config-path` will **not** be passed to risingwave process, so the defaults in code will be used. See discussions in https://github.com/risingwavelabs/risingwave/issues/6803.

Also did:

- Support `config-path` for `risedev compose`, which didn't work on #6678
- Refactored `expand_with_extra_info()` to make the code cleaner. 

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

None

## Refer to a related PR or issue link (optional)

Related #6678

Also closes #6803

### CODE REVIEW

**Review by xxchan — COMMENTED** [review]
So after some recent changes, current values in `config/risingwave.toml` are already the same as defaults in code, right? 🤔 

We should add some notes in the file and its README to be more clear that it's only a sample and should be kept in sync with the values in code.

File: src/risedevtool/src/config.rs [hunk 1]
```diff
@@ -69,62 +73,37 @@ impl ConfigExpander {
...
-                let v = id_expander.visit(v)?;
-                let mut provide_expander = ProvideExpander::new(&v)?;
-                let v = provide_expander.visit(v)?;
-                Ok::<_, anyhow::Error>((profile.clone(), v))
-            })
-            .try_collect()?;
+        let config_path = profile_map
+            .get(&Yaml::String("config-path".to_string()))
+            .and_then(|s| s.as_str())
+            .unwrap_or("src/config/risingwave.toml")
```
Comment (xxchan) [inline src/risedevtool/src/config.rs:87]: Well if the values in `config/risingwave.toml` is the same as defaults in code, and we want to use it as a sample, we can simply avoid using it completely? 🤔

File: src/cmd_all/src/playground.rs [hunk 2]
```diff
@@ -30,7 +30,7 @@ use tokio::fs::File;
 use tokio::io::AsyncReadExt;
 use tokio::signal;
 
-async fn load_risedev_config(profile: &str) -> Result<(Option<String>, Vec<ServiceConfig>)> {
+async fn load_risedev_config(profile: &str) -> Result<(String, Vec<ServiceConfig>)> {
```
[the remark this answers is not in the document]
Comment (fuyufjh) [reply src/cmd_all/src/playground.rs:33]: You mean `risingwave.toml` can be missing? In that case, `src/config/risingwave.toml` will be used. 

https://github.com/risingwavelabs/risingwave/pull/6814/files#diff-e1cea435bf1e8e5ef4692d74e955179ec3f3579c6e5e596d51849fe89c54b8acL66

File: src/risedevtool/src/config.rs (hunk 1 above)
Comment (fuyufjh) [reply src/risedevtool/src/config.rs:87]: Look good to me. So `risingwave.toml` will be useless at all.

Comment (fuyufjh) [reply src/risedevtool/src/config.rs:87]: Oops, that would require me to revert `String` back to `Option<String>` 🤣

Comment (fuyufjh) [reply src/risedevtool/src/config.rs]: Updated. Please take a look @xxchan

**Comment by fuyufjh** [comment]
@hzxa21 We are going to delete [risingwave.toml](https://github.com/risingwavelabs/risingwave/blob/88a8ebef9406583864b72f67126722208e781041/src/config/example.toml) and make the [config.rs](https://github.com/risingwavelabs/risingwave/blob/8a86cc4f3b35f7c69365516306cca926f7607370/src/common/src/config.rs#L389) the only source of defaults.

I just compared the default values in risingwave.toml and the items below have different default values in [risingwave.toml](https://github.com/risingwavelabs/risingwave/blob/88a8ebef9406583864b72f67126722208e781041/src/config/example.toml) and [config.rs]. Please take a look:

- block_cache_capacity_mb
- shared_buffer_capacity_mb
- meta_cache_capacity_mb
- compactor_memory_limit_mb

**Comment by hzxa21** [comment]
> @hzxa21 We are going to delete [risingwave.toml](https://github.com/risingwavelabs/risingwave/blob/88a8ebef9406583864b72f67126722208e781041/src/config/example.toml) and make the [config.rs](https://github.com/risingwavelabs/risingwave/blob/8a86cc4f3b35f7c69365516306cca926f7607370/src/common/src/config.rs#L389) the only source of defaults.
> 
> I just compared the default values in risingwave.toml and the items below have different default values in [risingwave.toml](https://github.com/risingwavelabs/risingwave/blob/88a8ebef9406583864b72f67126722208e781041/src/config/example.toml) and [config.rs](https://github.com/risingwavelabs/risingwave/blob/8a86cc4f3b35f7c69365516306cca926f7607370/src/common/src/config.rs#L389). Please take a look:
> 
> * block_cache_capacity_mb
> * shared_buffer_capacity_mb
> * meta_cache_capacity_mb
> * compactor_memory_limit_mb

Since the default configs are expected to be used in dev environment or playground, I think it is okay to keep the default values in config.rs (smaller than the ones set in risingwave.toml) for these configs. cc @Little-Wallace 

Btw, do we use customized toml config file For performance test? We should set the cache and buffer capacity based on the resource we have.

**Comment by BugenZhao** [comment]
> Btw, do we use customized toml config file For performance test? We should set the cache and buffer capacity based on the resource we have.

I guess so. 🤔 cc @huangjw806 We may need to take care of this and make sure non-default values are set.

**Comment by fuyufjh** [comment]
> > Btw, do we use customized toml config file For performance test? We should set the cache and buffer capacity based on the resource we have.
> 
> I guess so. 🤔 cc @huangjw806 We may need to take care of this and make sure non-default values are set.

cc. @jon-chuang @lmatz

File: src/config/README.md [hunk 3]
```diff
@@ -23,5 +23,3 @@ risedev:
     steps:
       - use: ...
 ```
-
```
Comment (BugenZhao) [inline src/config/README.md:26]: Should we also update the doc below? 

> `./risingwave.toml` is the default config used by `risedev` .

**Review by xxchan — APPROVED** [review]

**Comment by xxchan** [comment]
Fixed a problem for profiles without config-path

**Pull request merged**